File: custom_apps/org_logistics_control/org_logistics_control/public/js/venucopy_portal.js
(function initVenucopyPortal() {
  const kpiEl = document.getElementById("venucopy-kpis");
  const tableBody = document.querySelector("#venucopy-jobs-table tbody");
  const companyTag = document.getElementById("venucopy-company-tag");
  if (!kpiEl || !tableBody) return;

  fetch("/api/method/org_logistics_control.venucopy_portal.get_dashboard_data", {
    credentials: "same-origin",
  })
    .then((r) => r.json())
    .then((payload) => {
      const data = payload && payload.message ? payload.message : {};
      renderKpis(kpiEl, data.kpis || {});
      renderRows(tableBody, data.recent_jobs || []);
      if (companyTag) {
        companyTag.textContent = data.company ? `Company: ${data.company}` : "";
      }
    })
    .catch(() => {
      kpiEl.innerHTML = "<div class='venucopy-kpi'><div class='venucopy-kpi-label'>Error</div><div class='venucopy-kpi-value'>Unable to load</div></div>";
    });
})();

function renderKpis(container, kpis) {
  const blocks = [
    ["Total Jobs", kpis.jobs_total || 0],
    ["Payment Stage", kpis.jobs_payment || 0],
    ["Pending Delivery", kpis.jobs_pending_delivery || 0],
  ];

  container.innerHTML = blocks
    .map(
      ([label, value]) =>
        `<div class="venucopy-kpi"><div class="venucopy-kpi-label">${label}</div><div class="venucopy-kpi-value">${value}</div></div>`
    )
    .join("");
}

function renderRows(tbody, rows) {
  if (!rows.length) {
    tbody.innerHTML = "<tr><td colspan='7'>No jobs found</td></tr>";
    return;
  }

  tbody.innerHTML = rows
    .map((row) => {
      const ref = row.cargo_reference || "-";
      const customer = row.customer || row.supplier || "-";
      const margin = toMoney(row.total_margin);
      const url = `/app/logistics-job/${encodeURIComponent(row.name)}`;
      return `<tr>
        <td><a href="${url}">${escapeHtml(row.name)}</a></td>
        <td>${escapeHtml(ref)}</td>
        <td>${escapeHtml(customer)}</td>
        <td>${escapeHtml(row.order_status || "-")}</td>
        <td>${escapeHtml(row.payment_status || "-")}</td>
        <td>${escapeHtml(row.delivery_status || "-")}</td>
        <td>${margin}</td>
      </tr>`;
    })
    .join("");
}

function toMoney(value) {
  const n = Number(value || 0);
  return n.toLocaleString(undefined, { maximumFractionDigits: 2 });
}

function escapeHtml(value) {
  return String(value || "")
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#039;");
}

File: custom_apps/org_logistics_control/org_logistics_control/venucopy_portal.py
import frappe


@frappe.whitelist()
def get_dashboard_data():
    if frappe.session.user == "Guest":
        frappe.throw("Login required")

    company = (
        frappe.defaults.get_user_default("company")
        or frappe.defaults.get_user_default("Company")
        or frappe.db.get_single_value("Global Defaults", "default_company")
    )

    job_filters = {}
    if company and frappe.db.exists("DocType", "Logistics Job"):
        job_filters["company"] = company

    jobs_total = frappe.db.count("Logistics Job", filters=job_filters) if frappe.db.exists("DocType", "Logistics Job") else 0
    jobs_payment = (
        frappe.db.count("Logistics Job", filters={**job_filters, "order_status": "payment"})
        if frappe.db.exists("DocType", "Logistics Job")
        else 0
    )
    jobs_pending_delivery = (
        frappe.db.count("Logistics Job", filters={**job_filters, "delivery_status": "pending"})
        if frappe.db.exists("DocType", "Logistics Job")
        else 0
    )

    recent_jobs = []
    if frappe.db.exists("DocType", "Logistics Job"):
        recent_jobs = frappe.get_all(
            "Logistics Job",
            filters=job_filters,
            fields=[
                "name",
                "cargo_reference",
                "customer",
                "supplier",
                "order_status",
                "payment_status",
                "delivery_status",
                "total_sell_amount",
                "total_buy_amount",
                "total_margin",
                "modified",
            ],
            order_by="modified desc",
            limit=8,
        )

    return {
        "company": company or "",
        "kpis": {
            "jobs_total": jobs_total,
            "jobs_payment": jobs_payment,
            "jobs_pending_delivery": jobs_pending_delivery,
        },
        "recent_jobs": recent_jobs,
    }
